### PR TITLE
Recovery Tanker for carriers.

### PR DESCRIPTION
**Pull request by SnappyComebacks**
This adds support for a recovery tanker at a carriers.  It still needs a changelog, and depends on https://github.com/pydcs/dcs/pull/278.

This works on 6.x, but I haven't tested tankers that have already been scheduled on a carrier.  I will try that before deciding if this is a 6.x or 7.x feature.

@DanAlbert I wanted to get your thoughts on what I've done with `UnitMap` (adding the `dcs_group_id` field, and passing it into `pydcswaypointbuilder.py`.  Another option would be to pass in the group itself instead of just the id, with the hope of being able to use the waypoints to interpolate positions, but the carrier's waypoints are set pretty late (see `steam_into_wind`).

### CODE REVIEW

File: game/ato/flightplans/flightplanbuildertypes.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 from typing import Any, TYPE_CHECKING, Type
 
 from game.ato import FlightType
+from game.ato.flightplans.shiprecoverytanker import RecoveryTankerFlightPlan
```
Comment (DanAlbert) [inline game/ato/flightplans/flightplanbuildertypes.py:6]: Prefer `.shiprecoverytanker` (that's one of the tricks for making the circular import issues slightly less frequent, though I'm not sure it actually makes a difference in this case)

File: game/ato/flightplans/flightplanbuildertypes.py [hunk 2]
```diff
@@ -33,8 +35,13 @@ class FlightPlanBuilderTypes:
     @staticmethod
     def for_flight(flight: Flight) -> Type[IBuilder[Any, Any]]:
         if flight.flight_type is FlightType.REFUELING:
-            if flight.package.target.is_friendly(flight.squadron.player) or isinstance(
-                flight.package.target, FrontLine
+            target = flight.package.target
+            if target.is_friendly(flight.squadron.player) and isinstance(
+                target, Carrier
```
Comment (DanAlbert) [inline game/ato/flightplans/flightplanbuildertypes.py:40]: What about LHA? I think there's a parent that captures both of those.

File: game/ato/flightplans/shiprecoverytanker.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    def builder_type() -> Type[Builder]:
+        return Builder
+
+    @property
+    def tot_waypoint(self) -> FlightWaypoint:
+        return self.layout.recovery_ship
+
+    @property
+    def mission_departure_time(self) -> timedelta:
+        return timedelta(hours=2)
```
Comment (DanAlbert) [inline game/ato/flightplans/shiprecoverytanker.py:42]: This is absolute, not relative to the TOT. A recovery tanker with a TOT of 1:50 will stay on station for 10 minutes. A tanker with a TOT of 2:00 will not do anything. I think you just want `return self.patrol_end_time` here.

File: game/ato/flightplans/shiprecoverytanker.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        return self.tot + self.mission_departure_time
+
+    def tot_for_waypoint(self, waypoint: FlightWaypoint) -> timedelta | None:
+        if waypoint == self.tot_waypoint:
+            return self.tot
+        return None
+
+    def depart_time_for_waypoint(self, waypoint: FlightWaypoint) -> timedelta | None:
+        if waypoint == self.tot_waypoint:
+            return self.tot + self.mission_departure_time
```
Comment (DanAlbert) [inline game/ato/flightplans/shiprecoverytanker.py:59]: Here too. This should be `self.patrol_end_time`

File: game/ato/flightplans/shiprecoverytanker.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        return None
+
+
+class Builder(IBuilder[RecoveryTankerFlightPlan, RecoveryTankerLayout]):
+    def layout(self) -> RecoveryTankerLayout:
+
+        # TODO: Propagate the ship position.
```
Comment (DanAlbert) [inline game/ato/flightplans/shiprecoverytanker.py:66]: I don't know what this todo means. Could you expand the comment a bit? Do you mean predict the ship's position at the TOT rather than at mission start?

File: game/ato/flightplans/shiprecoverytanker.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        ship = self.package.target.position
+
+        builder = WaypointBuilder(self.flight, self.coalition)
+
+        recovery = builder.recovery_tanker(ship)
+
+        tanker_type = self.flight.unit_type
+        altitude = tanker_type.preferred_patrol_altitude
```
Comment (DanAlbert) [inline game/ato/flightplans/shiprecoverytanker.py:74]: Hmm. If the player tasks a KC-135 for this, the recovery tanker will be at 24k ft. Not useful. Not something they're likely to do either, but it doesn't seem like it's prevented. A couple options:

1. Make this a unique mission type so we can configure tasking allow/disallow separately, allow the S-3 (and whatever else is reasonable for a recovery tanker), but not the KC-135/KC-130, and make sure the recovery type tankers do it at a reasonable altitude (but that might make them worse for regular tanking missions).
2. Make this not aircraft specific, and low enough that you're not going to run out of fuel climbing back up to the gas station.

I'm not sure which behavior makes more sense, or if it should even do both (I have no idea if it's reasonable for a KC-130 to provide recovery tanking for a CVN). I do know that 24k ft is not a useful place to do it though :) WDYT?

File: game/ato/flightplans/shiprecoverytanker.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+        return RecoveryTankerLayout(
+            departure=builder.takeoff(self.flight.departure),
+            nav_to=builder.nav_path(self.flight.departure.position, ship, altitude),
```
Comment (DanAlbert) [inline game/ato/flightplans/shiprecoverytanker.py:78]: ... oh. It's actually the _cruise_ altitude? If that's all, mind `s/altitude/cruise_altitude/` and leave a comment like "we don't have per-aircraft cruise altitudes recorded yet, so just reuse the patrol altitude"?

File: game/missiongenerator/aircraft/waypoints/recoverytanker.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+from game.ato import FlightType
+from game.missiongenerator.missiondata import TankerInfo
+from game.utils import feet, knots
+from .pydcswaypointbuilder import PydcsWaypointBuilder
+
+
+class RecoveryTankerBuilder(PydcsWaypointBuilder):
+    def add_tasks(self, waypoint: MovingPoint) -> None:
+        if self.flight.flight_type == FlightType.REFUELING:
```
Comment (DanAlbert) [inline game/missiongenerator/aircraft/waypoints/recoverytanker.py:12]: Throw an exception if it isn't?

File: game/missiongenerator/aircraft/waypoints/recoverytanker.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        name = self.package.target.name
+        carrier_position = self.package.target.position
+        theater_objects = self.unit_map.theater_objects
+        for key, value in theater_objects.items():
+            # Check name and position in case there are multiple of same carrier.
+            if name in key and value.theater_unit.position == carrier_position:
+                theater_mapping = value
+                break
+        assert theater_mapping is not None
+        return theater_mapping.dcs_group_id
```
Comment (DanAlbert) [inline game/missiongenerator/aircraft/waypoints/recoverytanker.py:34]: I see what you mean. I think it'd be better to capture this in `MissionData` rather than in `UnitMap`. `UnitMap` is ugly because it has to do ugly things. You could add a `carrier_dcs_ids: dict[UUID, int]` that removes the name and position guesswork.

File: game/missiongenerator/tgogenerator.py [hunk 10]
```diff
@@ -147,7 +147,7 @@ def create_vehicle_group(
                 vehicle_unit.position = unit.position
                 vehicle_unit.heading = unit.position.heading.degrees
                 vehicle_group.add_unit(vehicle_unit)
-            self._register_theater_unit(unit, vehicle_group.units[-1])
+            self._register_theater_unit(vehicle_group.id, unit, vehicle_group.units[-1])
```
Comment (DanAlbert) [inline game/missiongenerator/tgogenerator.py:150]: And then I think you can revert the changes here and in `unitmap.py`

**Review by DanAlbert — COMMENTED** [review]
> This works on 6.x, but I haven't tested tankers that have already been scheduled on a carrier. I will try that before deciding if this is a 6.x or 7.x feature.

It looks safe, FWIW. An existing save will have the old waypoint type. It'll get the old behavior too, but anyone that wants the new feature on the same turn can just regenerate that one flight.

File: game/ato/flightplans/flightplanbuildertypes.py (hunk 2 above)
Comment (SnappyComebacks) [reply game/ato/flightplans/flightplanbuildertypes.py:40]: We presently do not support refueling at all on LHA in `6.0`.  I'm can add it if you think we should.

File: game/ato/flightplans/shiprecoverytanker.py (hunk 5 above)
Comment (SnappyComebacks) [reply game/ato/flightplans/shiprecoverytanker.py:66]: Yes, ideally we would pick the ship position at TOT.  In order to do that we need to get the ship's `group` down passed down here somehow.  I will update the comment.

File: game/ato/flightplans/shiprecoverytanker.py (hunk 3 above)
Comment (SnappyComebacks) [reply game/ato/flightplans/shiprecoverytanker.py:42]: Thanks for catching this.  I had given the `recovery` task a `Last Waypoint` of the final carrier waypoint, and needed _something_ here for `TankerInfo` used later on (which unless I missed it, the final time isn't displayed anywhere?).  Do you feel like giving me the tanker a final waypoint was the right call?  If so, ideally we'd need to get the carrier's waypoints and speed down to here.

Comment (SnappyComebacks) [reply game/ato/flightplans/shiprecoverytanker.py:42]: > I think you just want `return self.patrol_end_time` here.

I've just went and looked at this, if I change it to `patrol_end_time` the relationship becomes recursive.  I'll change the two methods so it's more in line with the other flight plans.

Comment (SnappyComebacks) [reply game/ato/flightplans/shiprecoverytanker.py:42]: > Thanks for catching this. I had given the `recovery` task a `Last Waypoint` of the final carrier waypoint, and needed _something_ here for `TankerInfo` used later on (which unless I missed it, the final time isn't displayed anywhere?). Do you feel like giving me the tanker a final waypoint was the right call? If so, ideally we'd need to get the carrier's waypoints and speed down to here.

I discovered I was wrong here with regards to TankerInfo, and `None` is valid.  I can get rid of `patrol_end_time` all together I think.

File: game/ato/flightplans/flightplanbuildertypes.py (hunk 2 above)
Comment (DanAlbert) [reply game/ato/flightplans/flightplanbuildertypes.py:40]: Seems like something we ought to support? I don't have strong feelings.

File: game/missiongenerator/aircraft/waypoints/recoverytanker.py (hunk 9 above)
Comment (SnappyComebacks) [reply game/missiongenerator/aircraft/waypoints/recoverytanker.py:34]: Do you mean `MissionData` or `MissionTarget` here?

Comment (DanAlbert) [reply game/missiongenerator/aircraft/waypoints/recoverytanker.py:34]: https://github.com/dcs-liberation/dcs_liberation/blob/develop/game/missiongenerator/missiondata.py
